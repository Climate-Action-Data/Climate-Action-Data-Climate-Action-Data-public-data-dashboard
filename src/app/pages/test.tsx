'use client'
import { type NextPage } from "next";

const Test: NextPage = () => {
    return (
        <div>
            <h1>Hello</h1>
            <p>data</p>
        </div>
    )
}

export default Test