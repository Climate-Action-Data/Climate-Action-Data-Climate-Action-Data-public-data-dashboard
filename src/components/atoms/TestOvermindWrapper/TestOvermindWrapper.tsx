import { createOvermind } from 'overmind'
import { Provider } from 'overmind-react'
import { config } from '@/overmind'
import { PropsWithChildren } from 'react'

const overmind = createOvermind(config)

export const TestOvermindWrapper = (props: PropsWithChildren) => {
  return <Provider value={overmind}>{props.children}</Provider>
}
