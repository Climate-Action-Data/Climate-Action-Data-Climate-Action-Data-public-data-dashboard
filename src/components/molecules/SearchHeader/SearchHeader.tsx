import { ChangeEvent, useEffect } from 'react'
import { usePathname, useRouter, useSearchParams } from 'next/navigation'
import { useTranslation } from 'react-i18next'
import { Box, Button, Flex, Heading, HStack, Input, InputGroup, InputRightElement, VStack } from '@chakra-ui/react'

import { extractPageFromUrl, extractTitleFromUrl } from '@/utils/TextConverter'

import { BreadCrumbs } from '@/components/atoms/BreadCrumbs/BreadCrumbs'
import { ESearchParams } from '@/@types/ProjectSearchResult'
import { useActions, useAppState } from '@/overmind'
import UnitFilterBar from '@/components/organisms/UnitFilterBar/UnitFilterBar'
import ProjectFilterBar from '@/components/organisms/ProjectFilterBar/ProjectFilterBar'

export const SearchHeader = () => {
  const { keywordSearch } = useAppState().searchFilters
  const { setKeywordSearch } = useActions().searchFilters
  const currentPath = usePathname()
  const currentTitle = extractTitleFromUrl(currentPath)
  const { t } = useTranslation(`search`)
  const searchParams = useSearchParams()
  const router = useRouter()
  const pattern = searchParams.get(ESearchParams.KEYWORD) ?? keywordSearch
  //TODO: refactor search to use state and not url params
  useEffect(() => {
    if (pattern) {
      if (searchParams.get(ESearchParams.KEYWORD) !== keywordSearch && keywordSearch !== ``) {
        const searchParams = new URLSearchParams()
        searchParams.append(ESearchParams.KEYWORD, keywordSearch)
        router.push(`/search/${extractPageFromUrl(currentPath)}?${searchParams}`)
      }
    }
  }, [])

  const handleOnSearch = () => {
    const searchParams = new URLSearchParams()
    searchParams.append(ESearchParams.KEYWORD, keywordSearch)
    setKeywordSearch(searchParams.get(ESearchParams.KEYWORD) ?? ``)
    router.push(`/search/${extractPageFromUrl(currentPath)}?${searchParams}`)
  }

  const handleOnChange = (event: ChangeEvent<HTMLInputElement>) => {
    setKeywordSearch(event.target.value)
  }

  const generateFilterBar = () => {
    if (currentPath === `/search/projects`) {
      return <ProjectFilterBar isResultsPage />
    } else if (currentPath === `/search/units`) {
      return <UnitFilterBar isResultsPage />
    }
  }

  return (
    <Flex id="headerReference" position="sticky" top={`56px`} zIndex="docked" padding="16px 24px" minH="184px" color="white" backgroundColor="gray.900" width="100%">
      <VStack alignItems="start" flex={1}>
        {currentPath && <BreadCrumbs items={[{ title: currentTitle, link: currentPath }]} />}
        <HStack w={`100%`} justifyContent={`space-between`}>
          <Box>
            <Heading>{currentTitle}</Heading>
          </Box>
          <Box>
            <InputGroup size="md">
              <Input pr="4.5rem" type={`text`} placeholder="Search" value={keywordSearch} onChange={handleOnChange} />
              <InputRightElement width="4.5rem">
                <Button colorScheme="white" variant="brandPrimary" h="1.75rem" size="sm" onClick={handleOnSearch} data-testid={`search-header-button`}>
                  {t(`search`)}
                </Button>
              </InputRightElement>
            </InputGroup>
          </Box>
        </HStack>
        {generateFilterBar()}
      </VStack>
    </Flex>
  )
}
