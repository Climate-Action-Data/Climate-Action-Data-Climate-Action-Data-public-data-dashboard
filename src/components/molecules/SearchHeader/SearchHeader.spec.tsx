import { fireEvent, render, screen } from '@testing-library/react'
import { SearchHeader } from '@/components/molecules/SearchHeader/SearchHeader'
import userEvent from '@testing-library/user-event'
import { TestOvermindWrapper } from '@/test/TestOvermindWrapper'
import { usePathname } from 'next/navigation'
import { MockData } from '@/test/TestOvermindMockData'

const mockPush = jest.fn()
jest.mock(`next/navigation`, () => ({
  ...jest.requireActual(`next/navigation`),
  useRouter: () => ({ push: mockPush }),
  usePathname: jest.fn(),
}))

beforeEach(() => {
  ;(usePathname as jest.Mock).mockReturnValue(`/search/projects`)
})

it(`renders correctly`, () => {
  const { container } = render(
    <TestOvermindWrapper>
      <SearchHeader />
    </TestOvermindWrapper>,
  )
  expect(container).toMatchSnapshot()
})

it(`renders correctly with search text`, () => {
  const mockData = {
    ...MockData.SEARCH_FILTER_VALUES,
    keywordSearch: `test`,
  }
  const { container } = render(
    <TestOvermindWrapper searchFilters={mockData}>
      <SearchHeader />
    </TestOvermindWrapper>,
  )
  expect(container).toMatchSnapshot()
})

it(`renders correctly for units path`, () => {
  ;(usePathname as jest.Mock).mockReturnValue(`/search/units`)

  const { container } = render(
    <TestOvermindWrapper>
      <SearchHeader />
    </TestOvermindWrapper>,
  )
  expect(container).toMatchSnapshot()
})

it(`renders correctly and clear the search values`, async () => {
  const { container } = render(
    <TestOvermindWrapper>
      <SearchHeader />
    </TestOvermindWrapper>,
  )
  await userEvent.clear(screen.getByRole(`textbox`))
  expect(container).toMatchSnapshot()
})

it(`renders correctly and clear the search values`, async () => {
  const { container } = render(
    <TestOvermindWrapper>
      <SearchHeader />
    </TestOvermindWrapper>,
  )
  const input = screen.getByTestId(`search-input-enter`)
  fireEvent.keyDown(input, {
    key: `Enter`,
    code: `Enter`,
    keyCode: 13,
    charCode: 13,
  })
  await userEvent.clear(screen.getByRole(`textbox`))
  expect(container).toMatchSnapshot()
})
const NUMBER_TIMES_ROUTE_CALLED = 2
const TIMEOUT = 100000

it(
  `renders correctly with no params, add a search value and submits a search`,
  async () => {
    const { container } = render(
      <TestOvermindWrapper>
        <SearchHeader />
      </TestOvermindWrapper>,
    )
    await userEvent.type(screen.getByRole(`textbox`), `hello`)
    await userEvent.click(screen.getByTestId(`search-header-button`))
    expect(mockPush).toBeCalledTimes(NUMBER_TIMES_ROUTE_CALLED)
    expect(container).toMatchSnapshot()
  },
  TIMEOUT,
)
