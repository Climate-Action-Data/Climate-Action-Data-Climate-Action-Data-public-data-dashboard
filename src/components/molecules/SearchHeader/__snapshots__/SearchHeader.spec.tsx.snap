// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    class="css-8y3igr"
    id="headerReference"
  >
    <div
      class="chakra-stack css-ryi8m5"
    >
      <div
        class="chakra-stack css-1jjq5p5"
      >
        <div
          class="css-0"
        >
          <h2
            class="chakra-heading css-0"
          />
        </div>
        <div
          class="css-0"
        >
          <div
            class="chakra-input__group css-bx0blc"
            data-group="true"
          >
            <input
              class="chakra-input css-1n0hj84"
              placeholder="Search"
              type="text"
              value=""
            />
            <div
              class="chakra-input__right-element css-vcnsx7"
            >
              <button
                class="chakra-button css-xsdlqf"
                data-testid="search-header-button"
                type="button"
              >
                Search
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly and clear the search values 1`] = `
<div>
  <div
    class="css-8y3igr"
    id="headerReference"
  >
    <div
      class="chakra-stack css-ryi8m5"
    >
      <div
        class="chakra-stack css-1jjq5p5"
      >
        <div
          class="css-0"
        >
          <h2
            class="chakra-heading css-0"
          />
        </div>
        <div
          class="css-0"
        >
          <div
            class="chakra-input__group css-bx0blc"
            data-group="true"
          >
            <input
              class="chakra-input css-1n0hj84"
              placeholder="Search"
              type="text"
              value=""
            />
            <div
              class="chakra-input__right-element css-vcnsx7"
            >
              <button
                class="chakra-button css-xsdlqf"
                data-testid="search-header-button"
                type="button"
              >
                Search
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly with no params, add a search value and submits a search 1`] = `
<div>
  <div
    class="css-8y3igr"
    id="headerReference"
  >
    <div
      class="chakra-stack css-ryi8m5"
    >
      <div
        class="chakra-stack css-1jjq5p5"
      >
        <div
          class="css-0"
        >
          <h2
            class="chakra-heading css-0"
          />
        </div>
        <div
          class="css-0"
        >
          <div
            class="chakra-input__group css-bx0blc"
            data-group="true"
          >
            <input
              class="chakra-input css-1n0hj84"
              placeholder="Search"
              type="text"
              value="hello"
            />
            <div
              class="chakra-input__right-element css-vcnsx7"
            >
              <button
                class="chakra-button css-xsdlqf"
                data-testid="search-header-button"
                type="button"
              >
                Search
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
